backend/spec/controller_agent_person_spec.rb

```diff
@@ -2,187 +2,523 @@
 require 'agent_spec_helper'
 
 describe 'Person agent controller' do
-
-  def create_person(opts = {})
-    create(:json_agent_person, opts)
-  end
-
-
-  it "lets you create a person and get them back" do
-    opts = {:names => [build(:json_name_person)]}
-
-    id = create_person(opts).id
-    expect(JSONModel(:agent_person).find(id).names.first['primary_name']).to eq(opts[:names][0]['primary_name'])
-  end
-
-
-  it "lets you update someone by adding contacts" do
-    id = create_person(:agent_contacts => nil).id
-
-    person = JSONModel(:agent_person).find(id)
-    [0, 1].each do |n|
-      opts = {:name => generate(:generic_name)}
-
-      person.agent_contacts << build(:json_agent_contact, opts)
-
-      person.save
-
-      expect(JSONModel(:agent_person).find(id).agent_contacts[n]['name']).to eq(opts[:name])
+  describe 'update action' do
+    describe 'adding contacts' do
+      context 'when person has no contacts' do
+        let(:agent_person) do
+          as_test_user("admin") do
+            create(:json_agent_person, :agent_contacts => nil)
+          end
+        end
+
+        context 'when user has the manage_agent_record permission on any repository' do
+          context 'when user has the view_agent_contact_record permission on any repository' do
+            before :each do
+              archivist = make_test_user("archivist")
+              Group[:group_code => 'repository-archivists', :repo_id => $repo_id].add_user(archivist)
+            end
+
+            it "persists the added contacts" do
+              as_test_user("archivist") do
+                person = JSONModel(:agent_person).find(agent_person.id)
+                first_contact_params = build(:json_agent_contact).to_hash
+                second_contact_params = build(:json_agent_contact).to_hash
+                person.agent_contacts = [first_contact_params, second_contact_params]
+                person.save
+
+                person = JSONModel(:agent_person).find(person.id)
+                expect(person.agent_contacts).to deep_include([first_contact_params, second_contact_params])
+              end
+            end
+
+            context 'when user does not have the view_agent_contact_record permission on any repository' do
+              before :each do
+                user = make_test_user("intern")
+                Group[:group_code => 'repository-advanced-data-entry', :repo_id => $repo_id].add_user(user)
+              end
+
+              it "ignores the added contacts" do
```

Comment:
@thimios it might be nice to alert the API user that they don't have permission to set or view the agent contacts. Is there an ArchivesSpace pattern for this?